### PR TITLE
Revised ES6 modules

### PR DESCRIPTION
**Pull request by ahejlsberg**
This PR addresses feedback on our design for ES6 module support (see #2242). The primary change is that `export default` and `export =` become separate constructs with different semantics:

* An `export default` declaration always declares an exported member named `default` and is always emitted as an assignment to `exports.default`. In other words, `export default` consistently has ES module semantics.
* An `export =` declaration, which substitutes a different entity to be exported *in place of the module itself*, is always emitted as an assignment to `module.exports`. It is an error to have other exports in a module that uses `export =`. This is the existing TypeScript behavior.
* A module that uses `export =` to export another module or a "module like" entity can be imported using the new ES6 constructs. In particular, the convenient destructuring imports can be used with such modules. The pattern of using `export =` to export another module is common in .d.ts files that provide a CommonJS/AMD view of an internal module (e.g. [angular.d.ts](https://github.com/borisyankov/DefinitelyTyped/blob/master/angularjs/angular.d.ts)).
* A module that uses `export =` to export a non-module entity in place of the module itself must be imported using the existing `import x = require("foo")` syntax as is the case today.

This approach has several advantages:

* It reduces the amount of magic used in mapping ES6 modules onto CommonJS/AMD.
* It gives us a better story for interop with existing declaration files.
* It allows circularly dependent modules to work in a consistent manner (including circularly dependent default exports). As has been pointed out here, the ability to have circular module references is actually one of the major advantages of the ES6 module design and it is important for us to preserve this attribute in down-level scenarios.

### CODE REVIEW

**Comment by ahejlsberg** [comment]
This PR also fixes #2451.

**Comment by ahejlsberg** [comment]
Also fixes #2383.

File: src/compiler/binder.ts [hunk 1]
```diff
@@ -123,7 +123,7 @@ module ts {
                 case SyntaxKind.ExportDeclaration:
                     return "__export";
                 case SyntaxKind.ExportAssignment:
-                    return "default";
+                    return (<ExportAssignment>node).isExportEquals ? "export=" : "default";
```
Comment (CyrusNajmabadi) [inline src/compiler/binder.ts]: this is a fairly odd name to give this entity.  Is it used elsewhere?  if so, we should put it in a common location so we don't have magic strings like this floating around.

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2835,17 +2876,25 @@ module ts {
             return result;
         }
 
-        function getExportsOfExternalModule(node: ImportDeclaration): Symbol[]{
+        function symbolsToArray(symbols: SymbolTable): Symbol[] {
+            let result: Symbol[] = [];
+            for (let id in symbols) {
+                if (!isReservedMemberName(id)) {
```
Comment (CyrusNajmabadi) [inline src/compiler/checker.ts]: Why aren't these included?

File: src/compiler/binder.ts (hunk 1 above)
Comment (ahejlsberg) [inline src/compiler/binder.ts]: It needs a name that can meaningfully be shown in error messages (such as duplicate symbol 'export=' when there is more than one). It is used in a few other places, but so are all the other ones here.

File: src/compiler/checker.ts (hunk 2 above)
Comment (ahejlsberg) [inline src/compiler/checker.ts]: We don't want the reserved member names (such as '__export') to show up in statement completion lists.

**Pull request merged**